docs/source/schema-directives.md

```diff
@@ -1,179 +1,581 @@
+```
+
+Points to notice in this example:
+
+* The `@deprecated` directive _follows_ the field that it pertains to (`oldField`), even though its syntax might remind you of "decorators" in other languages, which usually appear on the line above.
+* Directives are often _declared_ in the schema, as in this example, though it's up to the GraphQL server to enforce the argument types (`reason: String`) and locations (`FIELD_DEFINITION | ENUM_VALUE`) of the declaration.
+* The `@deprecated(reason: ...)` syntax is legal even without the `directive @deprecated ...` declaration.
+* Whether a directive has been declared or not, the GraphQL server is responsible for giving it meaning, or else it will be ignored after parsing, almost as if it was a comment.
+
+While only a few directives are [required](http://facebook.github.io/graphql/draft/#sec--skip) by the specification, the formal syntax of the GraphQL query and schema languages allows arbitrary user-defined directives to appear as modifiers following almost any kind of type, field, or argument. If the server knows how to interpret them, these directive annotations can be a powerful tool for preventing repetition, specifying extra behavior, enforcing additional type or value restrictions, and enabling static analysis.
+
+This document focuses on directives that appear in GraphQL _schemas_ (as opposed to queries) written in [Schema Definition Language](https://github.com/facebook/graphql/pull/90), or SDL for short. Specifically, the following sections will explain how custom directives can be implemented and used to modify the structure and behavior of a GraphQL schema in ways that would not be possible using SDL syntax alone.
+
+The possible applications of directive syntax are numerous: enforcing access permissions, formatting date strings, auto-generating resolver functions for a particular backend API, marking strings for internationalization, synthesizing globally unique object identifiers, specifying caching behavior, skipping or including or deprecating fields, and just about anything else you can imagine.
+
+## Implementing schema directives
+
+Since the GraphQL specification does not discuss any specific implementation strategy for directives, it's up to each GraphQL server framework to expose an API for implementing new directives.
+
+If you're using Apollo Server, you are also likely to be using the [`graphql-tools`](https://github.com/apollographql/graphql-tools) npm package, which provides a convenient yet powerful tool for implementing directive syntax: the [`SchemaDirectiveVisitor`](https://github.com/apollographql/graphql-tools/blob/wip-schema-directives/src/schemaVisitor.ts) class.
+
+To implement a schema directive using `SchemaDirectiveVisitor`, simply create a subclass of `SchemaDirectiveVisitor` that overrides one or more of the following visitor methods:
+
+* `visitSchema(schema: GraphQLSchema)`
+* `visitScalar(scalar: GraphQLScalarType)`
+* `visitObject(object: GraphQLObjectType)`
```

Comment:
It looks like these are intended to match the SDL types here: http://facebook.github.io/graphql/draft/#sec-Type-System.Directives

But it's a bit funny that they don't match the TypeScript/Flow types. Maybe we should link to the spec to explain where we got these names? IDK

Reply:
Yes, the names come directly from the `DirectiveLocation` enum values. I think they're pretty reasonable as method names, even if you don't know where they came from. I'd rather not throw `GraphQL` or `Type` into the names, though the relationship there is also fairly predictable. I'll add a link.